### PR TITLE
Drive StepKeyword parse and render from a single canonical keyword table (#511)

### LINKED ISSUE #511: Drive StepKeyword parse and render from a single canonical keyword table
## Summary

`StepKeyword::from_str` is a five-branch `else if` chain of
`eq_ignore_ascii_case` comparisons. The keyword literals (`given`, `when`,
`then`, `and`, `but`) are maintained here separately from the reverse direction
(`as_str`/display), so adding or renaming a keyword means editing two places
that can drift apart.

## Location

- `crates/rstest-bdd-patterns/src/keyword.rs:100` — the `from_str` if-else chain.

## Canonical-implementation mandate

- [ ] Introduce one canonical `const` mapping table (e.g.
      `const KEYWORDS: [(&str, StepKeyword); 5]`) as the single source of truth
      for the string ↔ keyword correspondence.
- [ ] Drive both `from_str` (case-insensitive lookup over the table) and the
      string-rendering side from this one table; remove the parallel literal
      lists.
- [ ] Document the table as the canonical keyword registry in
      `docs/developers-guide.md` (where to add a keyword, the round-trip
      contract), per the abstraction policy in `AGENTS.md`.

## Testing requirements

- [ ] **Property tests** (`proptest`): round-trip — for every keyword,
      `from_str(kw.as_str())` returns `kw`; and for an arbitrary ASCII-case
      permutation of each keyword string, `from_str` yields the same variant
      (case-insensitivity). An arbitrary non-keyword string returns the parse
      error.
- [ ] **Model checking**: not applicable — the domain is five variants and is
      covered exhaustively by the round-trip property plus example tests.
- [ ] **Snapshot assertions**: not applicable.

## Acceptance criteria

- [ ] One canonical keyword table; both parse and render derive from it.
- [ ] Developer's guide updated.
- [ ] Round-trip and case-insensitivity property tests added and passing.
- [ ] `make check-fmt`, `make lint`, `make test` pass.

## Context

Raised from a workspace refactoring audit. Tier 2 finding — addresses
parse/render literal drift, not just the conditional shape. Requested by
@leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #511

- Introduce a `keyword_table!` macro in `crates/rstest-bdd-patterns/src/keyword.rs` whose single invocation lists each `(rendering, variant)` pair exactly once and expands to both the `KEYWORDS` const table (driving the case-insensitive `from_str` lookup) and the exhaustive match inside the const `as_str`. The parallel literal lists are gone; parse and render cannot drift.
- A macro rather than a runtime table scan keeps `as_str` `const` and avoids slice indexing (denied by `clippy::indexing-slicing`).
- Add a property suite (`tests/keyword_props.rs`): round-trip for every keyword, invariance to arbitrary ASCII-case permutations and surrounding whitespace, and parse failure carrying the trimmed input for non-keyword strings. Model checking is not applicable per the issue (five-variant domain covered exhaustively).
- Record the canonical keyword registry and round-trip contract in `docs/developers-guide.md`.

## Testing

- `make check-fmt`, `make lint`, `make typecheck`, `make test` (1489 passed), `make markdownlint` all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Unify step keyword parsing and rendering around a single canonical keyword table and document its contract.

Bug Fixes:
- Ensure StepKeyword parsing and rendering stay consistent by driving both from one canonical keyword table and trimming input before lookup.

Enhancements:
- Introduce a keyword_table! macro that generates the canonical KEYWORDS mapping and StepKeyword::as_str implementation from a single list of keyword variants.
- Refactor StepKeyword::from_str to perform case-insensitive, whitespace-trimming lookup using the shared KEYWORDS table instead of hard-coded conditionals.

Documentation:
- Document the canonical step keyword registry, round-trip guarantees, and contribution guidelines in the developers guide.

Tests:
- Add property-based tests verifying keyword render/parse round-tripping, case- and whitespace-insensitive parsing, and error behavior for non-keyword strings.

### CODE REVIEW

File: crates/rstest-bdd-patterns/tests/keyword_props.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    fn round_trip_via_canonical_rendering(kw in keyword()) {
+        prop_assert_eq!(StepKeyword::from_str(kw.as_str()), Ok(kw));
+    }
+
+    /// Parsing is invariant to ASCII case permutations and surrounding
+    /// whitespace.
+    #[test]
+    fn parse_ignores_case_and_whitespace(
+        kw in keyword(),
+        mask in proptest::collection::vec(any::<bool>(), 0..5),
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-patterns/tests/keyword_props.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: docs/developers-guide.md [hunk 2]
```diff
@@ -845,6 +845,25 @@ feature-gated regression suite in
...
+
+`crates/rstest-bdd-patterns/src/keyword.rs` holds the `KEYWORDS` const table,
+the single source of truth for the string ↔ `StepKeyword` correspondence.
+`StepKeyword::as_str` (rendering) and `StepKeyword::from_str`
+(case-insensitive, whitespace-trimming parsing) both derive from this table;
+neither side carries its own literal list.
+
+- **Adding or renaming a keyword:** edit the `KEYWORDS` table (and the enum
+  variant) only. Do not add parallel literals to `as_str`, `from_str`, or
+  call-sites.
```
[the remark this answers is not in the document]
Comment (pandalump) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

**Comment by pandalump** [comment]
@coderabbitai review

File: crates/rstest-bdd-patterns/tests/keyword_props.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    #[test]
+    fn non_keyword_strings_fail_to_parse(input in "[a-zA-Z0-9_-]{0,12}") {
+        let is_keyword = ALL_KEYWORDS
+            .iter()
+            .any(|kw| input.trim().eq_ignore_ascii_case(kw.as_str()));
+        prop_assume!(!is_keyword);
+        prop_assert_eq!(
+            StepKeyword::from_str(&input),
+            Err(StepKeywordParseError(input.trim().to_string()))
+        );
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-patterns/tests/keyword_props.rs:100]: Done. `non_keyword_strings_fail_to_parse` now generates the core token and `leading`/`trailing` whitespace independently, builds a padded `input`, and asserts the error carries the trimmed `core`. When padding is non-empty the raw input differs from its trimmed form, so the trimming path in `from_str` is now genuinely exercised.

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai Re the **User-Facing Documentation** pre-merge warning: dismissing this as out of scope for this PR, with evidence.

- **No user-facing behaviour change.** #525 is an internal refactor: it drives `StepKeyword`'s `as_str` (render) and `from_str` (parse) from a single `keyword_table![...]` macro invocation instead of two hand-maintained literal lists. The observable behaviour is unchanged.
- **The case-insensitive/trimmed `from_str` is not a feature-file surface.** Its only non-test callers are internal (`crates/rstest-bdd-macros/src/step_keyword.rs` re-exports the `rstest-bdd-patterns` impl). Gherkin feature files are parsed by the `gherkin` crate (`StepType`), not by `StepKeyword::from_str`, so there is no user-facing "type `given` in your `.feature`" contract to document here.
- **The user-facing keywords are already documented.** `docs/gherkin-syntax.md` covers `Given`/`When`/`Then`/`And`/`But` for feature authors.
- **The canonical table is an internal implementation detail**, and this PR already documents it where it belongs — the developers' guide ("Canonical step-keyword table"), including the round-trip contract and the "edit the `keyword_table![...]` invocation" maintenance rule.

Adding a new users-guide section about an internal parsing utility would be an enhancement disproportionate to this internal-refactor PR and would risk asserting a user-facing contract (`from_str` on feature-file text) that does not exist. Happy to reconsider if you can point to a user-facing entry point that parses keyword text through `StepKeyword::from_str`.
